syntax = "proto2";

package Command;

enum Type {
  CMD_UNKNOWN = 0;
  CMD_GET_WINDOW_STATUS = 1;
  CMD_HIDE_WINDOW = 2;
  CMD_SHOW_WINDOW = 3;
  CMD_GET_STATUS = 4;
  CMD_TEST = 5;
  CMD_SET_VOLUME = 6;
  CMD_SEEK = 7;
  CMD_TOGGLE_SHUFFLE = 8;
  CMD_TOGGLE_REPEAT = 9;
  CMD_NEXT_TRACK = 10;
  CMD_PREV_TRACK = 11;
  CMD_PLAY = 12;
  CMD_PAUSE = 13;
  CMD_STOP = 14;
  CMD_FEATURE_BIG_COVER = 15;
  CMD_FEATURE_SHOW_CLOCK = 16;
  CMD_FAILURE = 17;
  CMD_FEATURE_SET_MAX_VOLUME = 18;
  CMD_SET_CLEAR_AUDIO = 19;
  CMD_SET_EQ_BANDS = 20;
  CMD_SET_EQ_PRESET = 21;
  CMD_SET_VPT = 22;
  CMD_SET_VPT_PRESET = 23;
  CMD_SET_DSEE = 24;
  CMD_SET_DCPHASE = 25;
  CMD_SET_DCPHASE_PRESET = 26;
  CMD_SET_VINYL = 27;
  CMD_SET_DIRECT_SOURCE = 28;
  CMD_SET_TONE_CONTROL_OR_EQ = 29;
  CMD_SET_TONE_CONTROL_VALUES = 30;
  CMD_SET_DSEE_CUST = 31;
  CMD_SET_DSEE_CUST_MODE = 32;
  CMD_SET_VINYL_MODE = 33;
}

enum ResponseCode {
  UNKNOWN = 0;
  FAIL = 1;
  OK = 2;
}

enum WindowVisible {
  VISIBILITY_UNKNOWN = 0;
  VISIBILITY_YES = 1;
  VISIBILITY_NO = 2;
}

message Command {
  required Type Type = 1;
  required ResponseCode Code = 2;

  oneof Msg {
    WindowStatus WindowStatus = 3;
    Status Status = 4;
    SetVolume SetVolume = 5;
    Seek Seek = 6;
    FeatureBigCover FeatureBigCover = 7;
    FeatureShowClock FeatureShowClock = 8;
    FeatureSetMaxVolume FeatureSetMaxVolume = 9;
    ClearAudio ClearAudio = 10;
    EqBands EqBands = 11;
    EqPreset EqPreset = 12;
    Vpt VPT = 13;
    VptPreset VPTPreset = 14;
    Dsee DSEE = 15;
    DCPhase DCPHASE = 16;
    DCPhasePreset DCPHASEPreset = 17;
    Vinyl Vinyl = 18;
    DirectSource DirectSource = 19;
    ToneControlValues ToneControlValues = 20;
    ToneControlOrEq ToneControlOrEq = 21;
    DseeCust DseeCust = 22;
    DseeCustMode DseeCustMode = 23;
    VinylMode VinylMode = 24;
  }
}

message FeatureBigCover {
  required bool enabled = 1;
}

message FeatureShowClock {
  required bool enabled = 1;
}

message FeatureSetMaxVolume {
  required bool enabled = 1;
}

message WindowStatus {
  required WindowVisible Visible = 1;
}

message SetVolume {
  required int32 valuePercent = 1;
  required bool relative = 2;
  required int32 relativeValue = 3;
}

message Seek {
  required int32 value = 1;
}

message Track {
  required int32 Track = 1;
  required string Artist = 2;
  required string Title = 3;
  required int32 Duration = 4;
  required bool Active = 5;
}

message Playlist {
  repeated Track Track = 1;
}

message Status {
  required string Codec = 1;
  required int32 Elapsed = 2;
  required int32 PlayState = 3;
  required bool HiRes = 4;
  required bool Shuffle = 5;
  required int32 Repeat = 6;
  // volume percentage
  required int32 Volume = 7;
  optional int32 BitRate = 8;
  optional float SampleRate = 9;
  optional int32 BitDepth = 10;
  required Playlist Playlist = 11;
  required int32 VolumeRaw = 12;
  required uint32 EntryID = 13;
}

message ClearAudio {
  required bool enabled = 1;
}

message EqBands {
  repeated double BandValue = 1;
}

message EqPreset {
  required int32 preset = 1;
}

message Vpt {
  required bool enabled = 1;
}

message VptPreset {
  required int32 preset = 1;
}

message Dsee {
  required bool enabled = 1;
}

message DCPhase {
  required bool enabled = 1;
}

message DCPhasePreset {
  required int32 preset = 1;
}

message Vinyl {
  required bool enabled = 1;
}

message DirectSource{
  required bool enabled = 1;
}

message ToneControlValues {
  required int32 low = 1;
  required int32 middle = 2;
  required int32 high = 3;
}

message ToneControlOrEq {
  required int32 eqId = 1;
}

message DseeCust {
  required bool enabled = 1;
}

message DseeCustMode {
  required int32 mode = 1;
}

message VinylMode {
  required int32 mode = 1;
}